[#9494] Consolidate GET course RESTFul API

ChooJeremy opened the pull request:
Part of #9494

**PR Checklist**

<!-- Remove this portion after you have made the checks. -->

**Outline of Solution**

<!-- Tell us how you solved the issue. -->
<!-- If there are things you want the reviewers to focus on, include them here as well. -->
<!-- This portion can be skipped if the fix is trivial. -->

GET /course will only serve the course related portions of the other APIs, and no other parts. To eliminate extra data, I've taken a look at what the UI shows for student and instructor and came up with the following:
GET /course.
If instructor:  input courseID, return CourseData for that course.

Ready for peer review.

Adoby7 commented on src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts:
```diff
@@ -174,23 +181,22 @@ export class InstructorSessionEditPageComponent extends InstructorSessionBasePag
-              }, (resp: ErrorMessageOutput) => {
-                this.statusMessageService.showErrorMessage(resp.error.message);
-              });
+    this.courseService.getCourse(this.courseId).subscribe((course: Course) => {
```
Use Response DTO here for Course?

Adoby7 commented on src/test/java/teammates/test/cases/webapi/GetCourseActionTest.java:
```diff
@@ -37,30 +38,49 @@ protected void testExecute() throws Exception {
-        assertEquals(courseAttributes.getId(), response.getCourseId());
-        assertEquals(courseAttributes.getName(), response.getCourseName());
-        assertEquals(courseAttributes.getTimeZone().getId(), response.getTimeZone());
+        assertEquals(courseAttributes.getId(), courseData.getCourseId());
```
Suggest to change `courseAttributes` to be `expectedCourse`

Adoby7 commented on src/main/java/teammates/ui/webapi/action/GetCourseAction.java:
```diff
@@ -17,7 +20,6 @@ protected AuthType getMinAuthLevel() {
     public void checkSpecificAccessControl() {
         String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
 
-        // TODO enable access for student
```
Should we enable access for student here? Perhaps need to ask core members

ChooJeremy replied:
Unfortunately I've spoken with Xiao Pu and he says that since student will return a list of courses that the student has access to, that focuses more on GET /courses instead of GET /course.

rrtheonlyone replied:
So as of now, a student cannot access this endpoint is it? 

Right now, Gatekeeper will throw an error if no instructor is found from the id.

ChooJeremy replied:
Yup! The action has been moved to /courses.

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/GetCourseAction.java:
```diff
@@ -17,7 +20,6 @@ protected AuthType getMinAuthLevel() {
 
-        // TODO enable access for student
         gateKeeper.verifyAccessible(
                 logic.getInstructorForGoogleId(courseId, userInfo.getId()),
```
Potential NPE here if userInfo is `null` (unregistered user)

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/GetCourseAction.java:
```diff
@@ -26,9 +28,10 @@ public void checkSpecificAccessControl() {
         CourseAttributes courseAttributes = logic.getCourse(courseId);
-        return new JsonResult(new CourseInfo.CourseResponse(courseAttributes));
+        if (courseAttributes == null) {
+            throw new EntityNotFoundException(new EntityDoesNotExistException("No course with id: " + courseId));
```
Perhaps we should stick to the convention of returning a 404 `JsonResult` here? Thats what I have been doing in my other PRs.

rrtheonlyone commented on src/test/java/teammates/test/cases/webapi/GetCourseActionTest.java:
```diff
@@ -29,38 +30,57 @@ protected String getRequestMethod() {
 
     @Test
     @Override
     protected void testAccessControl() throws Exception {
```
What if a `student` or `unregistered user` calls this endpoint?

ChooJeremy replied:
The `verifyOnlyInstructorsOfTheSameCourseCanAccess` call (at the end of the method) checks for student and unregistered users.

ChooJeremy commented:
Ready for review

junming403 commented:
`If student: no input. output array of CourseData (only ID, name and timeZone provided) - this is based off functionality from GetStudentCourseAction`

Where is this part reflected?

junming403 commented on src/test/java/teammates/test/cases/webapi/GetCourseActionTest.java:
```diff
@@ -29,38 +30,61 @@ protected String getRequestMethod() {
         verifyHttpParameterFailure();
 
-        ______TS("typical success case");
+        ______TS("typical success case for instructor");
```
Instead of using `____TS()`. It's better to separate tests into different sub tests and annotate them with `@Test`. Since you already did it for one of the function, let's do it for all of them?

ChooJeremy commented:
> 
> 
> `If student: no input. output array of CourseData (only ID, name and timeZone provided) - this is based off functionality from GetStudentCourseAction`
> 
> Where is this part reflected?

That part's functionality has been moved to get /courses, as this API should only return a single course object (instead of multiple, like an array)

ChooJeremy commented:
Sorry, it turns out that student support is required in this endpoint after all, from STUDENT_COURSE (GET). It has been updated to support students. The latest 2 commits contains all the changes since the last LGTM, please take a look :)

Adoby7 commented on src/test/java/teammates/test/cases/webapi/GetCourseActionTest.java:
```diff
@@ -97,15 +97,28 @@ protected void testAccessControl() throws Exception {
         };
 
-        verifyOnlyInstructorsOfTheSameCourseCanAccess(submissionParams);
+        verifyInaccessibleWithoutLogin(submissionParams);
```
You are advised to put this change in "Updated tests to support students" for a better practice of organizing commit messages

ChooJeremy replied:
Should it? I kind of disagree... this change is relevant to the change in access control for GetCourseAction, so the test should be updated in the same commit that contains that change?

ChooJeremy commented:
The [build passes](https://travis-ci.org/TEAMMATES/teammates/builds/506776134), but does not appear to be correctly being reported in the UI here.

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetCourseAction.java:
```diff
@@ -16,20 +19,42 @@ protected AuthType getMinAuthLevel() {
+        }
+
+        if (userInfo.isStudent) {
+            if (!hasStudentJoinedCourse(courseId)) {
```
This `if` check is not necessary I think. `gateKeeper.verifyAccessible(logic.getStudentForGoogleId(courseId, userInfo.id), course);` will do null check.

xpdavid commented on src/web/services/course.service.ts:
```diff
@@ -15,6 +15,16 @@ export class CourseService {
   }
 
+  /**
+   * Get course data as an instructor by calling API.
```
```suggestion
   * Get course data by calling API.
```

ChooJeremy commented on src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts:
```diff
@@ -149,10 +150,16 @@ export class InstructorSessionEditPageComponent extends InstructorSessionBasePag
+              statusMessageService: StatusMessageService,
+              navigationService: NavigationService,
+              feedbackSessionsService: FeedbackSessionsService,
+              feedbackQuestionsService: FeedbackQuestionsService,
```
Hmm... making this change causes:
```
ERROR in src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts(60,14): error TS2415: Class 'InstructorSessionEditPageComponent' incorrectly extends base class 'InstructorSessionBasePageComponent'.
  Property 'feedbackQuestionsService' is private in type 'InstructorSessionEditPageComponent' but not in type 'InstructorSessionBasePageComponent'.
```
for each of the properties changing to private. If we change the access modifier of the base class, we'll need to modify the access modifier for all the other classes that extend from it as well (instructor sessions page component etc.)

xpdavid replied:
Opps. Merged but not seen this comment :P

wkurniawan07 replied:
Then it's `protected`; still not the default which is `public`
